minitask2/akhmad/behaviour.py

```diff
@@ -0,0 +1,196 @@
+                self.publish_command(left_turn_cmd)
+                print("Obstacles on front and right - turning left")
+            elif min(right) <= 0.5 and min(front) >= 0.5:
+                self.publish_command(move_cmd)
+                print("No obstacles in front but on right - move forward")
+            elif min(right) >= 0.5 and min(front) <= 0.5:
+                self.publish_command(left_turn_cmd)
+                print("No obstacles on right but in front - turning left")
+            elif min(right) >= 0.5 and min(front) >= 0.5:
+                if min(right_front) <= 0.5:
+                    self.publish_command(right_turn_cmd)
+                else:
+                    self.publish_command(move_cmd)
+                print("No obstacles on right and in front - sliding right")
+
+    def obstacle_avoidance(self):
+        # continue_right and continue_left is used to
+        # keep turning in same direction as robot started
+        # otherwise it might get stuck in corners
+        if continue_right:
+            self.turn_right()
+        elif continue_left:
+            self.turn_left()
+        elif all(i >= 0.5 for i in left_front):
+            self.turn_left()
+        elif all(i >= 0.5 for i in right_front):
+            self.turn_right()
+        elif all(i >= 0.5 for i in left_back):
+            self.turn_left()
+        elif all(i >= 0.5 for i in right_back):
+            self.turn_right()
+        else:
+            self.turn_left()
+
+    def random_walk(self):
+        rospy.loginfo("Random walk")
+        global save_position, continue_right, continue_left
+        global last_pose_theta, last_pose_x, last_pose_y
+
+        continue_right = False
+        continue_left = False
+
+        # move forward 3 m and turn by random angle
+        if sqrt((self.current_pose_x - last_pose_x) ** 2 +
+                (self.current_pose_y - last_pose_y) ** 2) < 3:
+            self.publish_command(move_cmd)
+            save_position = False
+        else:
+            self.publish_command(Twist())
+            save_position = True
+            for x in range(0, random.randint(0, 120)):
+                last_pose_theta = self.current_pose_theta
+                last_pose_x = self.current_pose_x
+                last_pose_y = self.current_pose_y
+
+                self.publish_command(left_turn_cmd)
+                self.rate.sleep()
+
+    def scan_callback(self, scan):
+        global front, right_front, right_back, back, left_back, left_front, right, left
+        global last_pose_theta
+        global last_pose_x
+        global last_pose_y
+
+        front = scan.ranges[330:] + scan.ranges[0:30]
+        left_front = scan.ranges[30:90]
+        left_back = scan.ranges[90:150]
+        left = scan.ranges[60:120]
+        back = scan.ranges[150:210]
+        right_back = scan.ranges[210:270]
+        right_front = scan.ranges[270:330]
+        right = scan.ranges[240:300]
+
+        if self.behaviour == 1:
```

Comment:
This `1` defo needs to be a constant. As it is, this line means nothing if you don't read the entire rest of the code. If you changed it to be `if self.behaviour == BEHAVIOUR_FOLLOW_WALL:` it would be obvious what its doing by itself.

Reply:
I agree